### ISSUE TITLE
[object_detection] Training Error: ssd_mobilenet_v2. Number of feature maps is expected to equal the length of `num_anchors_per_location`

### ISSUE THREAD

**Issue by luqiang6q** [tensorflow/models#4903]
# I want to use two feature maps for predicting
Then I change the file ssd_mobilenet_v2.config.
as follows:
    anchor_generator {
      ssd_anchor_generator {
        num_layers: 2
        min_scale: 0.1
        max_scale: 0.3
        aspect_ratios: 1.0
      }
    }

# Another change in ssd_mobilenet_v2_feature_extractor.py
      feature_map_layout = {
        'from_layer': ['layer_15/expansion_output', 'layer_19'],
        'layer_depth': [-1, -1],
        'use_depthwise': self._use_depthwise,
        'use_explicit_padding': self._use_explicit_padding,
    }      

Then the error occurred:

Traceback (most recent call last):
  File "train.py", line 184, in <module>
    tf.app.run()
  File "/home/lq/anaconda3/lib/python3.6/site-packages/tensorflow/python/platform/app.py", line 126, in run
    _sys.exit(main(argv))
  File "train.py", line 180, in main
    graph_hook_fn=graph_rewriter_fn)
  File "/home/lq/tf-models/research/object_detection/trainer.py", line 285, in train
    clones = model_deploy.create_clones(deploy_config, model_fn, [input_queue])
  File "/home/lq/anaconda3/lib/python3.6/site-packages/slim-0.1-py3.6.egg/deployment/model_deploy.py", line 193, in create_clones
  File "/home/lq/tf-models/research/object_detection/trainer.py", line 198, in _create_losses
    prediction_dict = detection_model.predict(images, true_image_shapes)
  File "/home/lq/tf-models/research/object_detection/meta_architectures/ssd_meta_arch.py", line 402, in predict
    im_width=image_shape[2]))
  File "/home/lq/tf-models/research/object_detection/core/anchor_generator.py", line 100, in generate
    raise ValueError('Number of feature maps is expected to equal the length '
ValueError: Number of feature maps is expected to equal the length of `num_anchors_per_location`.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
What is the top-level directory of the model you are using
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by mmkkyy** [comment]
Hi @luqiang6q,
I face the same problem when I try to speed up the model. 
Did you solve it?

**Comment by MetaPeak** [comment]
@luqiang6q @mmkkyy   When you write a new feature extractor network, suppose  it use 3 feature maps to do multibox location&classification, then you need to change the num_layers from  6 to 3 in the   config file.
 **The number of feature maps used to do multibox location and classification must much the groups of anchors**   
```
    anchor_generator {
      ssd_anchor_generator {
        num_layers: 6  change it to 3
        min_scale: 0.2
        max_scale: 0.95
        aspect_ratios: 1.0
        aspect_ratios: 2.0
        aspect_ratios: 0.5
        aspect_ratios: 3.0
        aspect_ratios: 0.3333
      }
    }
```

**Comment by LewsTherin511** [comment]
Did you maybe solve the issue? Because I'm having the same problem with _ssd_mobilenet_v1_. I added a '_Conv2d_5_pointwise_1 to the feature map and increased the _num_layers_ from 6 to 7. but when running model_main.py I get the _Number of feature maps is expected to equal the length of `num_anchors_per_location`._ message.

**Comment by yhliang2018** [comment]
Hi, are you using the Keras application model of mobilenet_v1/v2, or the slim version?

**Comment by LewsTherin511** [comment]
Hi, thanks for the reply!
If I'm not misunderstanding, I'm NOT using the slim version.

I tested it on two configurations:
(1)
OS Platform and Distribution: Ubuntu 18.04 LTS
TensorFlow installed from: pip binary 
TensorFlow version: 1.12.0

and (2)
OS Platform and Distribution: Ubuntu 18.04 LTS
TensorFlow installed from: pip binary 
TensorFlow version: tensorflow-gpu 1.13.1

and getting the same issue on both.

I'm trying to re-train an SSD model to detect one class of custom objects (guitars), by using the ssd_mobilenet_v1_coco model.
I wanted to add an extra feature map (Conv2d_5_pointwise) to the ones already present, thus having a total of 7 feature maps. So, I modified the "models/ssd_mobilenet_v1_feature_extractor.py" this way:

` feature_map_layout = {
        'from_layer': ['Conv2d_5_pointwise','Conv2d_11_pointwise', 'Conv2d_13_pointwise', '', '',
                       '', ''][:self._num_layers],
        'layer_depth': [-1, -1, -1, 512, 256, 256, 128][:self._num_layers],
        'use_explicit_padding': self._use_explicit_padding,
        'use_depthwise': self._use_depthwise,
    }`

And, accordingly, I changed num_layers into the config file to 7 too.

`   anchor_generator {
      ssd_anchor_generator {
        num_layers: 7
        min_scale: 0.2
        max_scale: 0.95
        aspect_ratios: 1.0
        aspect_ratios: 2.0
        aspect_ratios: 0.5
        aspect_ratios: 3.0
        aspect_ratios: 0.3333
      }
    }`

as well as the other necessary lines (path of tfrecord, label_map, number of classes and model checkpoint to train from).
However, when trying to train the model with main_model.py, I get the error message

`File "~/projects/tf_models/research/object_detection/core/anchor_generator.py", line 105, in generate
    raise ValueError('Number of feature maps is expected to equal the length '
ValueError: Number of feature maps is expected to equal the length of `num_anchors_per_location`.`

**Comment by LewsTherin511** [comment]
> Hi, are you using the Keras application model of mobilenet_v1/v2, or the slim version?

Hi @yhliang2018 , can I maybe provide more info? I asked on StackOverflow as well, but I didn't get any answer yet and I'm still stuck on this thing.

**Comment by yhliang2018** [comment]
@LewsTherin511 Yes, it would be helpful if you could provide more details, or even a code snippet to reproduce this error.

**Comment by LewsTherin511** [comment]
Hi @yhliang2018 , I am not sure as to what you would need to know, sorry if the message turns out to be of little use!

As I said, I'm trying to retrain one of the provided models (ssd_mobilenet_v1_coco) on a custom dataset.
This is my configuration:

OS Platform and Distribution: Ubuntu 18.04 LTS
TensorFlow installed from: pip binary
TensorFlow version: 1.12.0

I downloaded some 1K images from OpenIMage and derived the corresponding TFrecord files.
I re-trained the model by using the 'model_main.py' script, exported the inference graph and used it to detect the custom object (guitars) in new images/videos. Everything works fine, but I've been experiencing issues when detecting small objects.

So, I wanted to try and add a feature map to see if the detection improves.
To do so, I modified the file "models/ssd_mobilenet_v1_feature_extractor.py" this way:

```
feature_map_layout = {
        'from_layer': ['Conv2d_5_pointwise','Conv2d_11_pointwise', 'Conv2d_13_pointwise', '', '',
                       '', ''][:self._num_layers],
        'layer_depth': [-1, -1, -1, 512, 256, 256, 128][:self._num_layers],
        'use_explicit_padding': self._use_explicit_padding,
        'use_depthwise': self._use_depthwise,
    }
```
i.e., I added 'Conv2d_5_pointwise' in 'from_layer', and a '-1' at the beginning of 'layer_depth'.
Accordingly, I modified the .config I'm using for the transfer learning:

 ```
  anchor_generator {
      ssd_anchor_generator {
        num_layers: 7
        min_scale: 0.2
        max_scale: 0.95
        aspect_ratios: 1.0
        aspect_ratios: 2.0
        aspect_ratios: 0.5
        aspect_ratios: 3.0
        aspect_ratios: 0.3333
      }
    }
```

by increasing num_layers to 7.
Now I'm doing the same procedure as before, so basically I'm running:

`python model_main.py --pipeline_config_path=training_config/ssd_mobilenet_v1_coco.config --model_dir=training_logs/test/ --sample_1_of_n_eval_examples=1 --alsologtostderr
`

and I'm getting the error:

```
File "/home/lews/PycharmProjects/tf_models/research/object_detection/core/anchor_generator.py", line 105, in generate
    raise ValueError('Number of feature maps is expected to equal the length '
ValueError: Number of feature maps is expected to equal the length of `num_anchors_per_location`.
```

I tried a couple of other things:

- I modified one of the layers already present to 'Conv2d_5_pointwise', instead of adding it at the beginning, so that I have:

```
'from_layer': ['Conv2d_5_pointwise', 'Conv2d_13_pointwise', '', '',
                       '', ''][:self._num_layers],
        'layer_depth': [-1, -1, 512, 256, 256, 128][:self._num_layers],
```
with

```
        num_layers: 6
```
and it works.

- I deleted one layer, bringing the total to 5, so I used:
```

        'from_layer': ['Conv2d_13_pointwise', '', '',
                       '', ''][:self._num_layers],
        'layer_depth': [-1, 512, 256, 256, 128][:self._num_layers],
```

with

```
        num_layers: 5
```

and it works again.


- I added 2 feature maps, so:

```
        'from_layer': ['Conv2d_5_pointwise', 'Conv2d_5_pointwise', 'Conv2d_11_pointwise', 'Conv2d_13_pointwise', '', '',
                       '', ''][:self._num_layers],
        'layer_depth': [-1, -1, -1, -1, 512, 256, 256, 128][:self._num_layers],
```
with

```
        num_layers: 8
```
and it doesn't work (same error message).

So, long story short, it seems to be giving me this error whenever I try to use more than 6 feature maps, even if I select the right number of layers in the .config file.

**Comment by LewsTherin511** [comment]
Ok, figured it out.

Simply, I had to modify another parameter (***num_layers***) in the constructor of the SSDMobileNetV1FeatureExtractor class:

    def __init__(self,
               is_training,
               depth_multiplier,
               min_depth,
               pad_to_multiple,
               conv_hyperparams_fn,
               reuse_weights=None,
               use_explicit_padding=False,
               use_depthwise=False,

               num_layers=7,    <--- HERE

               override_base_feature_extractor_hyperparams=False):

to match the new number of feature maps.

**Comment by timatom** [comment]
While I know this issue has been closed for awhile, I wanted to throw a solution of mine into this thread.

I was getting the same "Training Error: ssd_mobilenet_v2. Number of feature maps is expected to equal the length of num_anchors_per_location" error when trying to update the anchor generator parameters in the `pipeline.config` file. Namely, I was getting this error when changing the `ssd_anchor_generator.num_layers` parameter from 6 to 3. This is common practice when trying to reduce the number of anchor (or prior) boxes that are fed into an NMS at inference time, i.e., gives faster inference times.

Instead of updating the `SSDMobileNetV2FeatureExtractor` class, though, I created a new `feature_extractor.num_layers` parameter (in the same `pipeline.config` file) with the same value as the `ssd_anchor_generator.num_layers` parameter. This way I'm not fiddling inside the TF codebase (which should always be a last resort).

Let me know if this comment isn't clear enough and I'll update it accordingly. Really wish these things were talked about more in TF's OD API docs.
